lib/ash/changeset/changeset.ex

```diff
@@ -1094,50 +1094,58 @@ defmodule Ash.Changeset do
   end
 
   @doc false
+  # Returns either an appropriate expression for an atomic condition or a value
+  # indicated that the condition cannot be handled atomically.
+  #
+  # Validation logic matches on failure. So, for example, `present(:field)` is
+  # going to _match_ when `:field` is `nil`. However, when applying this logic
+  # to a `where` condition, the opposite is desired. The end result is kinda
+  # ugly because it can end up reading like "not is not equal to" but
+  # ultimately produces the correct results.
+  @spec atomic_condition([{module(), keyword()}], Ash.Changeset.t(), map()) ::
+          {:atomic, Ash.Expr.t() | boolean()} | {:not_atomic, String.t()}
   def atomic_condition(where, changeset, context) do
     Enum.reduce_while(where, {:atomic, true}, fn {module, validation_opts},
-                                                 {:atomic, condition} ->
+                                                 {:atomic, condition_expr} ->
       case module.atomic(
              changeset,
              validation_opts,
              struct(Ash.Resource.Validation.Context, context)
            ) do
         :ok ->
-          {:cont, {:atomic, condition}}
-
-        {:error, _} ->
-          {:cont, {:atomic, false}}
-
-        [{:atomic, _, expr, _as_error} | rest] ->
-          exprs = [expr | Enum.map(rest, &elem(&1, 2))]
-
-          new_expr =
-            Enum.reduce(exprs, condition, fn expr, condition ->
-              if condition == true do
-                expr
-              else
-                expr(^condition and ^expr)
-              end
-            end)
-
-          {:cont, {:atomic, new_expr}}
+          {:cont, {:atomic, condition_expr}}
 
         {:atomic, _, expr, _as_error} ->
-          new_expr =
-            if condition == true do
-              expr
-            else
-              expr(^condition and ^expr)
-            end
-
-          {:cont, {:atomic, new_expr}}
+          {:cont, {:atomic, atomic_condition_expr(condition_expr, expr)}}
 
-        {:not_atomic, reason} ->
-          {:halt, {:not_atomic, reason}}
+        {:error, _} ->
+          # Error from the validator, so the validations should just fail with
+          # a `false` expression.
+          {:halt, {:atomic, false}}
+
+        {:not_atomic, _reason} = not_atomic ->
+          {:halt, not_atomic}
+
+        atomic_conditions when is_list(atomic_conditions) ->
+          atomic_conditions
+          |> Stream.map(fn {:atomic, _, expr, _as_error} -> expr end)
+          |> Enum.reduce(condition_expr, &atomic_condition_expr(&2, &1))
+          |> then(&{:cont, {:atomic, &1}})
```

Comment:
This was a combination of better DRY and what I felt was safer matching. This ensures every item in the list conforms to the correct (as much as it can) format and I think it reads better too.

I don't actually have a test to check this path. What's an example of a situation where this would be returned from `atomic/3` on a validation?

Reply:
I tried writing an explicit test for this but after tossing `IO.inspect/3` in the mix, it appears this is getting checked by the tests and everything is indeed hunky dory.

Reply:
You could ensure that you are testing this by either 
1: using `Ash.bulk_update(a_query, :the_action, %{...input}, strategy: :atomic)`
2: testing with an update action that has `require_atomic?: true` (the default for update actions)

Reply:
You're already doing #2 so yes, hunky dory indeed ❤️